File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/graphql/Cursors.java
package com.zappymart.adapters.graphql;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

public final class Cursors {

    private Cursors() {
    }

    public static String encode(String identifier) {
        return Base64.getUrlEncoder().withoutPadding()
                .encodeToString(identifier.getBytes(StandardCharsets.UTF_8));
    }

    public static String decode(String cursor) {
        if (cursor == null) {
            return null;
        }
        try {
            return new String(Base64.getUrlDecoder().decode(cursor), StandardCharsets.UTF_8);
        } catch (IllegalArgumentException notACursor) {
            return cursor;
        }
    }
}

File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/graphql/Cursors.kt
package nl.zappymart.adapters.graphql

import java.util.Base64

object Cursors {

    fun of(id: String): String = Base64.getUrlEncoder().withoutPadding().encodeToString(id.toByteArray())

    fun idOf(cursor: String?): String? {
        if (cursor.isNullOrBlank()) {
            return null
        }
        return try {
            String(Base64.getUrlDecoder().decode(cursor))
        } catch (malformed: IllegalArgumentException) {
            null
        }
    }
}
